Ia 2068 json data store

quang-le opened the pull request:
In order to receive data calculated outside of Iaso, we need an endpoint that lets users define a slug and associates json data to it

Related JIRA tickets : IA-2068

## Self proofreading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [ ] Are my typescript files well typed
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [X] My migrations file are included
- [X] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- Add `/api/datastore/` endpoint. The endpoint supports GET, POST, PUT and DELETE
- Add separate read and write permissions for the endpoint. The permissions are not visible from the front-end but can be attributed through the Django admin
- Add JsonDataStore entry to the Django admin
- GET: results returned are account specific (you only see data stores for the account of the user you're logged as)
- POST: the body must include `key` (the slug) and `data` (the json). The key must be unique for the account
- PUT: must include all fields. Will return a 400 if you try to update a key with another that already exists for your account.
- DELETE: well, it deletes.

## How to test

- go to /api/datastore/
- try creating a few stores by POSTing
- try to access a specific store directly via `/api/datastore/{slug}`
- try to update a store with PUT
- try to DELETE and see if the list updates

## Print screen / video

https://user-images.githubusercontent.com/38907762/233620678-76d8dcfd-abb8-4a8d-99d0-f35273fa22b1.mov

## Notes

I used `slugify` to sanitize the slugs, I don't know if it enough

quang-le committed:
IA-2068: datastore api v1

quang-le committed:
black

quang-le committed:
IA-2068: add delete support to endpoint

quang-le committed:
IA-2068: add json data store to django admin

quang-le committed:
black

quang-le committed:
IA-2048: add tests

quang-le committed:
IA-2048: add slug sanitization

quang-le committed:
IA-2048: add tests

quang-le committed:
IA-2068: hide permissions from front-end

quang-le committed:
black

quang-le committed:
IA-2068: ignore my py error

quang-le committed:
black

mathvdh commented on iaso/api/data_store.py:
```diff
@@ -0,0 +1,81 @@
+
+        if request.method == "GET":
+            can_get = (
+                request.user and any(request.user.has_perm(perm) for perm in read_perms) or request.user.is_superuser
```
Tu peux simplifier ça :
```
read_perm = "menupermissions.iaso_datastore_read"
```

Puis : 
```
can_get = (request.user and request.user.has_perm(read_perm)) or request.user.is_superuser
```

Idem pour `write_perm` ...

mathvdh replied:
Aussi faut pas rajouter un `request.user.is_authenticated` ?

quang-le replied:
Pas faux. J'avais repris le code avec les tuples ailleurs en me disant que ce serait plus simple si on voulait donner accès avec d'autres permissions plus tard

mathvdh commented on iaso/api/data_store.py:
```diff
@@ -0,0 +1,81 @@
+            JsonDataStore.objects.filter(account=self.context["request"].user.iaso_profile.account)
+        )
+        method = self.context["request"].method
+        key_already_exists = request_key in [store.slug for store in stores_for_account]
```
Tu peux faire plus optimisé avec un truc du genre : 

```python
key_already_exists  = JsonDataStore.objects.filter(account=self.context["request"].user.iaso_profile.account, slug=request_key).exists()
```

La tu loaded toutes les keys à chaque fois ...

mathvdh reviewed (changes requested):
A few minor changes but otherwise it's good !

quang-le committed:
IA-2068: code review

- add check on use auth in api
- simply perm check
- optimize check on slug unicity for account

quang-le committed:
IA-2068: fix typo related bug

mathvdh reviewed (approved):
👍🏼
